File: algorithm/mergeSort.js
function merge(input, low, mid, high){
	var i=low, j=mid+1,k=0;
	var temp = new Array(high-low+1);

	while(i<=mid && j<=high){
		if(input[i]<input[j]){
			temp[k++] = input[i++];
		}else{
			temp[k++] = input[j++];
		}
	}
	
	while(i<=mid) temp[k++] = input[i++];
	
	while(j<=high) temp[k++] = input[j++];

	for(i=low,k=0;i<=high;k++,i++){
		input[i] = temp[k];
	}
}

function sort(input, n){ //input元素个数
	
	var size = 1, low, mid, high;

	while(size <= n-1){
		low = 0;
		while(low+size <= n-1){
			mid  = low + size - 1;
			high = mid + size;  
			if(high > n-1){
				high = n-1;
			}

			merge(input, low, mid, high);
			low = high + 1;
		}
		size *= 2;
	}
}

var input = [3,1,2,7,4,5,8,6];
var n = input.length;

console.log(input);
sort(input, n);
console.log(input);
input = null;

File: algorithm/mergeSort.cc
#include <iostream>
#include <vector>

using namespace std;

void Merge(vector<int> &vv,int low,int mid,int high){
    int i=low,j=mid+1;
    int k=0;
    vector<int> v2(high-low+1);
    while(i<=low && j<=high){
        if(vv[i]<=vv[j])
            v2[k++] = vv[i];
        else
            v2[k++] = vv[j];
    }

    while(i<=low)
        v2[k++] = vv[i];
    while(j<=high)
        v2[k++] = vv[j];

    for(i=low,k=0;k<v2.size();++i,++k)
        vv[i] = v2[k];


}

void MergeSort(vector<int> & vv){
    int intvl=1,low,mid,high;
    int len = vv.size();

    while(intvl<=len-1){
        low = 0;
        while(high<=len-1){
            mid = low+intvl-1;
            high = mid + intvl;
            Merge(vv,low,mid,high);

        }
        intvl *= 2;

    }
}

int main(){

    vector<int> vv={2,3,1,4,8,6,7,5,9};
    MergeSort(vv);
    for(int i=0;i<vv.size();++i)
        cout << vv[i] << " ";
    cout << endl;

    return 0;
}
